src/sink/gcs.rs

```diff
@@ -0,0 +1,345 @@
+        mut event: Event,
+    ) -> ResultVec {
+        let remote = if let Some(remote) = &self.remote {
+            remote
+        } else {
+            self.remote = Some(auth::json_api_client(&self.config.pem, &HeaderMap::new()).await?);
+            let remote = self.remote.as_ref().ok_or("Client error!")?;
+            remote
+            // TODO - Qos checks
+        };
+        for value in event.value_iter() {
+            let command = parse_command(value)?;
+            let command_str = command.to_string();
+            let response = match command {
+                StorageCommand::Fetch(bucket_name, object_name) => {
+                    storage::get_object(&remote, &bucket_name, &object_name).await?
+                }
+                StorageCommand::ListBuckets(project_id) => {
+                    storage::list_buckets(&remote, &project_id).await?
+                }
+                StorageCommand::ListObjects(bucket_name) => {
+                    storage::list_objects(&remote, &bucket_name).await?
+                }
+                StorageCommand::Add(bucket_name, object, file_path) => {
+                    storage::add_object(&remote, &bucket_name, &object, &file_path).await?
+                }
+                StorageCommand::Remove(bucket_name, object) => {
+                    storage::delete_object(&remote, &bucket_name, &object).await?
+                }
+                StorageCommand::Create(project_id, bucket_name) => {
+                    storage::create_bucket(&remote, &project_id, &bucket_name).await?
+                }
+                StorageCommand::Delete(bucket_name) => {
+                    storage::delete_bucket(&remote, &bucket_name).await?
+                }
+                StorageCommand::Download(bucket_name, object_name) => {
+                    storage::download_object(&remote, &bucket_name, &object_name).await?
+                }
+                StorageCommand::Unknown => {
+                    warn!(
+                        "Unknown Google Cloud Storage command: `{}` attempted",
+                        command.to_string()
+                    );
+
+                    return Err(format!(
+                        "Unknown Google Cloud Storage command: `{}` attempted",
+                        command.to_string()
+                    )
+                    .into());
+                }
+            };
+            let response: Value = json!({
+                "cmd": command_str,
+                "data": response
+            })
+            .into();
+            if self.is_linked {
+                if let Some(reply_channel) = &self.reply_channel {
+                    reply_channel
+                        .send(sink::Reply::Response(
+                            OUT,
+                            Event {
+                                id: EventId::new(1, 2, 3),
+                                data: LineValue::new(vec![], |_| ValueAndMeta::from(response)),
```

Comment:
We'll need to include the corresponding correlation metadata here. the `es` sink has a good example for that